tmt/base.py

```diff
@@ -2313,7 +2388,8 @@ def __init__(self,
                  *,
                  id_: Optional[str] = None,
                  tree: Optional[Tree] = None,
-                 context: Optional['tmt.cli.Context'] = None) -> None:
+                 context: Optional['tmt.cli.Context'] = None,
+                 logger: tmt.log.Logger) -> None:
```

Comment:
Looking at this I realized: Having the `logger` as the last parameter somehow seems more natural to me... Not a big deal though, just wanted to share my feelings ;-)

Reply:
Hmmm... Looking more and more through the code, the feeling gets stronger and stronger...

Reply:
As long as we'd be dealing with keyword-only arguments, it's perfectly fine. I learned to state it as the first one, but no strong opinions here. If tmt's way would be "thou shalt declare logger as the last parameter", be it.

Reply:
> If tmt's way would be "thou shalt declare logger as the last parameter", be it.

Thanks. Let's do it this way. The important parameters first and `logger` as the last one. When/if adding new parameters in the future `logger: tmt.log.Logger) -> None` will stay nicely untouched as the last line together with the return value.

Reply:
One thing to resolve before I dive into moving parameters around: consider methods and functions with optional parameters:

```python
def environment_files_to_dict(
        env_files: Iterable[str],
        root: Optional[str] = None) -> EnvironmentType:
```

With logger as the always-first parameter, it looks like this, nice & tidy:

```python
def environment_files_to_dict(
        logger: tmt.log.Logger,
        env_files: Iterable[str],
        root: Optional[str] = None) -> EnvironmentType:
```

In the current state, `logger` cannot be truly last, Python would not allow that because `root` is a keyword argument with a default value, so we have two options:

```python
# last-but-not-so-last - this I find especially confusing
def environment_files_to_dict(
        env_files: Iterable[str],
        logger: tmt.log.Logger,
        root: Optional[str] = None) -> EnvironmentType:

# Turn the function into the keyword-only mode, then the logger can be last while other parameters keep their default values
def environment_files_to_dict(
        *,
        env_files: Iterable[str],
        root: Optional[str] = None,
        logger: tmt.log.Logger,) -> EnvironmentType:
```

Reply:
I think turning functions with optional arguments to keyword-only mode makes sense and does not hurt much. It's a fair price for having the essential parameters at the beginning.

Reply:
Ok, I'll add a commit tomorrow moving logger to the last position.

(`logger` *is* essential too O:)

Reply:
> (`logger` _is_ essential too O:)

Yeah, sure, agreed, no doubt about it :)

Reply:
First attempt, e3fa4f9f1e26099c13e35e6706b5599296c35f11

Reply:
Thanks much! Changes look good.